our-paxos/paxos.py

```diff
@@ -28,31 +28,146 @@ def parse_cfg(cfgpath):
+
+    # put size as last element
+    msg = msg << 16 | len(loc)
+    nbytes += 2
+
+    #encode
+    return msg.to_bytes(nbytes, 'big', signed=True)
+
+
+def paxos_decode(msg_bin):
+    """Decode a paxos binary message received from a network socket 
+    into a loc (list of chunks), maintaining message order
+
+    msg_bin -- encoded paxos message in binary
+    """
+    msg = int.from_bytes(msg_bin, byteorder='big', signed=True)
+    loc = [] #list of chunks
+    # original list size encoded in last chunk of the message
+    num_of_chunks = msg & 2**16 -1 #bitmask last 2 bytes
+    msg = msg >> 16
+
+    # extract 2bytes chunks in the message
+    for i in range(num_of_chunks):
+        loc.insert(0, msg & 2**16 -1) #bitmask last 2 bytes
+        msg = msg >> 16
+    
+    return loc
+
+
+
 def acceptor(config, id):
     print ('-> acceptor', id)
-    state = {}
+    # dictionary of acceptor states where
+    # { paxos_instance: state }
+    paxos_instances = {} 
+    
     r = mcast_receiver(config['acceptors'])
     s = mcast_sender()
     while True:
+        # init
+        init_state = {"rnd": 0, "v-rnd": 0, "v-val": 0} # acceptor state
+        # recv to a large buffer
         msg = r.recv(2**16)
-        # fake acceptor! just forwards messages to the learner
-        if id == 1:
-            # print "acceptor: sending %s to learners" % (msg)
-            s.sendto(msg, config['learners'])
+        # extract the instance and phase
+        loc = paxos_decode(msg)
+        instance = loc[0]
+        phase = loc[1]
+        
+        # set the initial state if it's the first time we start
+        # this paxos instance
+        if instance not in paxos_instances.keys():
+            paxos_instances[instance] = init_state
+
+        # let's change the current state
+        state = paxos_instances[instance]
+        if phase == 1:
+            # received phase 1A msg from proposer
+            if loc[2] > state['rnd']:
+                # update state and paxos_
+                state['rnd'] = loc[2]
+                print("i: {} p: {}, state: {}".format(instance, phase, state))
+                # encode and send phase 1B to proposer
+                msg = paxos_encode([instance, phase, state['rnd'], state['v-rnd'], state['v-val']])
+                s.sendto(msg, config['proposers'])
+        elif phase == 2:
+            # received phase 2A msg from proposer
+            # loc[2] = c-rnd, loc[3] = c-val
+            if loc[2] >= state['rnd']:
+                #state['rnd'] = loc[2] # not in the slides, but it makes sense (?)
+                state['v-rnd'] = loc[2]
+                state['v-val'] = loc[3]
+                print("i: {} p: {}, state: {}".format(instance, phase, state))
+
+                # encode and send phase 2B to proposer
+                msg = paxos_encode([instance, phase, state['v-rnd'], state['v-val']])
+                s.sendto(msg, config['learners'])
```

Comment:
Same thing as in the previous comment, if I remember correctly we send 2B in any case